File: src/cloud/storage.ts
import { randomBytes } from 'crypto';

export interface UploadOptions {
  contentType?: string;
  public?: boolean;
  metadata?: Record<string, string>;
  cacheControl?: string;
  encryptionKey?: string;
}

export interface UploadResult {
  url: string;
  path: string;
  size: number;
  contentType: string;
  etag: string;
  uploadedAt: Date;
}

export interface StorageObject {
  path: string;
  size: number;
  contentType: string;
  etag: string;
  uploadedAt: Date;
  lastModified: Date;
  metadata: Record<string, string>;
}

interface Bucket {
  name: string;
  region: string;
  createdAt: Date;
  objects: Map<string, StorageObject>;
  blobStore: Map<string, Buffer>;
}

class StorageError extends Error {
  constructor(message: string, public readonly code: string) {
    super(message);
    this.name = 'StorageError';
  }
}

function generateEtag(): string {
  return `"${randomBytes(16).toString('hex')}"`;
}

export class StorageService {
  private buckets: Map<string, Bucket> = new Map();

  constructor(private readonly baseUrl: string, private readonly apiKey: string) {
    this.createInternalBucket();
  }

  async upload(
    data: Buffer,
    path: string,
    options?: UploadOptions
  ): Promise<UploadResult> {
    const bucketName = this.resolveBucket(path);
    const bucket = this.getOrCreateBucket(bucketName);
    const objectPath = this.normalizePath(path);
    const now = new Date();

    const object: StorageObject = {
      path: objectPath,
      size: data.length,
      contentType: options?.contentType ?? 'application/octet-stream',
      etag: generateEtag(),
      uploadedAt: now,
      lastModified: now,
      metadata: options?.metadata ?? {},
    };

    bucket.objects.set(objectPath, object);
    bucket.blobStore.set(objectPath, data);

    return {
      url: `${this.baseUrl}/storage/v1/${objectPath}`,
      path: objectPath,
      size: data.length,
      contentType: object.contentType,
      etag: object.etag,
      uploadedAt: now,
    };
  }

  async uploadFromFile(
    filePath: string,
    destPath: string,
    options?: UploadOptions
  ): Promise<UploadResult> {
    const { readFileSync } = await import('fs');
    const data = readFileSync(filePath);
    return this.upload(data, destPath, options);
  }

  async download(path: string): Promise<Buffer> {
    const bucketName = this.resolveBucket(path);
    const bucket = this.getBucket(bucketName);
    const objectPath = this.normalizePath(path);

    const blob = bucket.blobStore.get(objectPath);
    if (!blob) {
      throw new StorageError(`Object not found: ${path}`, 'NOT_FOUND');
    }
    return blob;
  }

  async downloadToFile(path: string, destPath: string): Promise<void> {
    const data = await this.download(path);
    const { writeFile } = await import('fs/promises');
    await writeFile(destPath, data);
  }

  async delete(path: string): Promise<void> {
    const bucketName = this.resolveBucket(path);
    const bucket = this.getBucket(bucketName);
    const objectPath = this.normalizePath(path);

    if (!bucket.objects.has(objectPath)) {
      throw new StorageError(`Object not found: ${path}`, 'NOT_FOUND');
    }

    bucket.objects.delete(objectPath);
    bucket.blobStore.delete(objectPath);
  }

  async deleteBatch(paths: string[]): Promise<number> {
    let count = 0;
    for (const path of paths) {
      try {
        await this.delete(path);
        count++;
      } catch {
        continue;
      }
    }
    return count;
  }

  async exists(path: string): Promise<boolean> {
    try {
      const bucketName = this.resolveBucket(path);
      const bucket = this.getBucket(bucketName);
      const objectPath = this.normalizePath(path);
      return bucket.objects.has(objectPath);
    } catch {
      return false;
    }
  }

  async getMetadata(path: string): Promise<StorageObject> {
    const bucketName = this.resolveBucket(path);
    const bucket = this.getBucket(bucketName);
    const objectPath = this.normalizePath(path);

    const object = bucket.objects.get(objectPath);
    if (!object) {
      throw new StorageError(`Object not found: ${path}`, 'NOT_FOUND');
    }
    return { ...object, metadata: { ...object.metadata } };
  }

  async list(prefix?: string, recursive?: boolean): Promise<StorageObject[]> {
    const results: StorageObject[] = [];
    for (const [, bucket] of this.buckets) {
      for (const [, object] of bucket.objects) {
        if (prefix && !object.path.startsWith(prefix)) {
          continue;
        }
        if (!recursive && prefix) {
          const remaining = object.path.slice(prefix.length);
          if (remaining.includes('/')) {
            continue;
          }
        }
        results.push({ ...object, metadata: { ...object.metadata } });
      }
    }
    return results;
  }

  async copy(source: string, dest: string): Promise<string> {
    const data = await this.download(source);
    const metadata = await this.getMetadata(source);
    await this.upload(data, dest, {
      contentType: metadata.contentType,
      metadata: metadata.metadata,
    });
    return dest;
  }

  async move(source: string, dest: string): Promise<string> {
    await this.copy(source, dest);
    await this.delete(source);
    return dest;
  }

  async getSignedUrl(path: string, expiresIn = 3600): Promise<string> {
    const objectPath = this.normalizePath(path);
    const expiry = new Date(Date.now() + expiresIn * 1000).toISOString();
    const signature = randomBytes(16).toString('hex');
    return `${this.baseUrl}/storage/v1/${objectPath}?token=${signature}&expires=${expiry}`;
  }

  getPublicUrl(path: string): string {
    const objectPath = this.normalizePath(path);
    return `${this.baseUrl}/storage/v1/public/${objectPath}`;
  }

  async createBucket(name: string, region?: string): Promise<void> {
    if (this.buckets.has(name)) {
      throw new StorageError(`Bucket already exists: ${name}`, 'ALREADY_EXISTS');
    }
    this.buckets.set(name, {
      name,
      region: region ?? 'us-east',
      createdAt: new Date(),
      objects: new Map(),
      blobStore: new Map(),
    });
  }

  async deleteBucket(name: string): Promise<void> {
    if (!this.buckets.has(name)) {
      throw new StorageError(`Bucket not found: ${name}`, 'NOT_FOUND');
    }
    this.buckets.delete(name);
  }

  async listBuckets(): Promise<string[]> {
    return [...this.buckets.keys()];
  }

  private createInternalBucket(): void {
    const now = new Date();
    this.buckets.set('default', {
      name: 'default',
      region: 'us-east',
      createdAt: now,
      objects: new Map(),
      blobStore: new Map(),
    });
  }

  private resolveBucket(path: string): string {
    const parts = path.replace(/^\/+/, '').split('/');
    return parts.length > 1 ? parts[0] : 'default';
  }

  private normalizePath(path: string): string {
    return path.replace(/^\/+/, '');
  }

  private getOrCreateBucket(name: string): Bucket {
    if (!this.buckets.has(name)) {
      this.buckets.set(name, {
        name,
        region: 'us-east',
        createdAt: new Date(),
        objects: new Map(),
        blobStore: new Map(),
      });
    }
    return this.buckets.get(name)!;
  }

  private getBucket(name: string): Bucket {
    const bucket = this.buckets.get(name);
    if (!bucket) {
      throw new StorageError(`Bucket not found: ${name}`, 'NOT_FOUND');
    }
    return bucket;
  }
}

export { StorageError };

File: zoya/cloud/storage.py
from dataclasses import dataclass, field
from typing import Optional, Dict, List, Any
import secrets
import time
import os
import mimetypes


@dataclass
class UploadOptions:
    content_type: str = "application/octet-stream"
    public: bool = False
    metadata: Dict[str, str] = field(default_factory=dict)
    cache_control: Optional[str] = None
    encryption_key: Optional[str] = None


@dataclass
class UploadResult:
    url: str
    path: str
    size: int
    content_type: str
    etag: str
    uploaded_at: float


@dataclass
class StorageObject:
    path: str
    size: int
    content_type: str
    etag: str
    uploaded_at: float
    last_modified: float
    metadata: Dict[str, str] = field(default_factory=dict)


class StorageError(Exception):
    def __init__(self, message: str, code: str = "STORAGE_ERROR"):
        self.code = code
        super().__init__(message)


def _generate_etag() -> str:
    return secrets.token_hex(16)


def _normalize_path(path: str) -> str:
    parts = [p for p in path.replace("\\", "/").split("/") if p and p != "."]
    stack: List[str] = []
    for part in parts:
        if part == "..":
            if stack:
                stack.pop()
        else:
            stack.append(part)
    return "/".join(stack)


class StorageService:
    def __init__(self, base_url: str, api_key: str):
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._buckets: Dict[str, Dict[str, StorageObject]] = {}
        self._bucket_data: Dict[str, Dict[str, bytes]] = {}
        self._bucket_metadata: Dict[str, Dict[str, Any]] = {}
        self._create_bucket_internal("default")

    def _get_upload_url(self, bucket: str, path: str) -> str:
        return f"{self._base_url}/storage/v1/{bucket}/{path}"

    def _get_public_url(self, bucket: str, path: str) -> str:
        return f"{self._base_url}/storage/v1/public/{bucket}/{path}"

    def _create_bucket_internal(self, name: str, region: str = "us-east"):
        if name not in self._buckets:
            self._buckets[name] = {}
            self._bucket_data[name] = {}
            self._bucket_metadata[name] = {"region": region, "created_at": time.time()}

    def _get_bucket_or_raise(self, name: str = "default") -> str:
        if name not in self._buckets:
            raise StorageError(f"Bucket '{name}' does not exist", "BUCKET_NOT_FOUND")
        return name

    def upload(
        self,
        data: bytes,
        path: str,
        options: Optional[UploadOptions] = None,
        bucket: str = "default",
    ) -> UploadResult:
        self._get_bucket_or_raise(bucket)
        options = options or UploadOptions()
        normalized = _normalize_path(path)
        content_type = options.content_type
        if content_type == "application/octet-stream":
            guessed, _ = mimetypes.guess_type(normalized)
            if guessed:
                content_type = guessed

        etag = _generate_etag()
        now = time.time()

        obj = StorageObject(
            path=normalized,
            size=len(data),
            content_type=content_type,
            etag=etag,
            uploaded_at=now,
            last_modified=now,
            metadata=dict(options.metadata),
        )
        self._buckets[bucket][normalized] = obj
        self._bucket_data[bucket][normalized] = data

        url = self._get_public_url(bucket, normalized) if options.public else self._get_upload_url(bucket, normalized)
        return UploadResult(
            url=url,
            path=normalized,
            size=len(data),
            content_type=content_type,
            etag=etag,
            uploaded_at=now,
        )

    def upload_from_file(
        self,
        file_path: str,
        dest_path: str,
        options: Optional[UploadOptions] = None,
        bucket: str = "default",
    ) -> UploadResult:
        if not os.path.isfile(file_path):
            raise StorageError(f"File not found: {file_path}", "FILE_NOT_FOUND")
        with open(file_path, "rb") as f:
            data = f.read()
        inferred_type, _ = mimetypes.guess_type(file_path)
        opts = UploadOptions(**(options.__dict__ if options else {}))
        if inferred_type and opts.content_type == "application/octet-stream":
            opts.content_type = inferred_type
        return self.upload(data, dest_path, opts, bucket)

    def download(self, path: str, bucket: str = "default") -> bytes:
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        if normalized not in self._bucket_data[bucket]:
            raise StorageError(f"Object not found: {normalized}", "OBJECT_NOT_FOUND")
        return self._bucket_data[bucket][normalized]

    def download_to_file(self, path: str, dest_path: str, bucket: str = "default"):
        data = self.download(path, bucket)
        os.makedirs(os.path.dirname(os.path.abspath(dest_path)), exist_ok=True)
        with open(dest_path, "wb") as f:
            f.write(data)

    def delete(self, path: str, bucket: str = "default"):
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        self._buckets[bucket].pop(normalized, None)
        self._bucket_data[bucket].pop(normalized, None)

    def delete_batch(self, paths: List[str], bucket: str = "default") -> int:
        self._get_bucket_or_raise(bucket)
        count = 0
        for p in paths:
            normalized = _normalize_path(p)
            if normalized in self._buckets[bucket]:
                self._buckets[bucket].pop(normalized)
                self._bucket_data[bucket].pop(normalized)
                count += 1
        return count

    def exists(self, path: str, bucket: str = "default") -> bool:
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        return normalized in self._buckets[bucket]

    def get_metadata(self, path: str, bucket: str = "default") -> StorageObject:
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        obj = self._buckets[bucket].get(normalized)
        if obj is None:
            raise StorageError(f"Object not found: {normalized}", "OBJECT_NOT_FOUND")
        return obj

    def list(
        self,
        prefix: Optional[str] = None,
        recursive: bool = False,
        bucket: str = "default",
    ) -> List[StorageObject]:
        self._get_bucket_or_raise(bucket)
        results: List[StorageObject] = []
        for obj_path, obj in self._buckets[bucket].items():
            if prefix and not obj_path.startswith(_normalize_path(prefix)):
                continue
            if not recursive and prefix:
                remainder = obj_path[len(_normalize_path(prefix)):].lstrip("/")
                if "/" in remainder:
                    continue
            results.append(obj)
        return sorted(results, key=lambda o: o.path)

    def copy(self, source: str, dest: str, bucket: str = "default") -> str:
        self._get_bucket_or_raise(bucket)
        src_normalized = _normalize_path(source)
        dest_normalized = _normalize_path(dest)
        obj = self._buckets[bucket].get(src_normalized)
        if obj is None:
            raise StorageError(f"Source not found: {src_normalized}", "OBJECT_NOT_FOUND")
        data = self._bucket_data[bucket][src_normalized]
        now = time.time()
        new_obj = StorageObject(
            path=dest_normalized,
            size=obj.size,
            content_type=obj.content_type,
            etag=_generate_etag(),
            uploaded_at=now,
            last_modified=now,
            metadata=dict(obj.metadata),
        )
        self._buckets[bucket][dest_normalized] = new_obj
        self._bucket_data[bucket][dest_normalized] = data
        return self._get_upload_url(bucket, dest_normalized)

    def move(self, source: str, dest: str, bucket: str = "default") -> str:
        url = self.copy(source, dest, bucket)
        self.delete(source, bucket)
        return url

    def get_signed_url(
        self,
        path: str,
        expires_in: int = 3600,
        bucket: str = "default",
    ) -> str:
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        if normalized not in self._buckets[bucket]:
            raise StorageError(f"Object not found: {normalized}", "OBJECT_NOT_FOUND")
        expiry = int(time.time()) + expires_in
        token = secrets.token_urlsafe(16)
        return f"{self._base_url}/storage/v1/signed/{bucket}/{normalized}?expires={expiry}&token={token}"

    def get_public_url(self, path: str, bucket: str = "default") -> str:
        self._get_bucket_or_raise(bucket)
        normalized = _normalize_path(path)
        if normalized not in self._buckets[bucket]:
            raise StorageError(f"Object not found: {normalized}", "OBJECT_NOT_FOUND")
        return self._get_public_url(bucket, normalized)

    def create_bucket(self, name: str, region: str = "us-east"):
        if name in self._buckets:
            raise StorageError(f"Bucket already exists: {name}", "BUCKET_EXISTS")
        self._create_bucket_internal(name, region)

    def delete_bucket(self, name: str):
        if name == "default":
            raise StorageError("Cannot delete default bucket", "BUCKET_PROTECTED")
        if name not in self._buckets:
            raise StorageError(f"Bucket not found: {name}", "BUCKET_NOT_FOUND")
        del self._buckets[name]
        del self._bucket_data[name]
        del self._bucket_metadata[name]

    def list_buckets(self) -> List[str]:
        return list(self._buckets.keys())
